kernel/rros/observable.rs

```diff
@@ -0,0 +1,970 @@
+    nfr = unsafe {
+        Arc::get_mut_unchecked(observer)
+            .pending_list
+            .pop_front()
+            .unwrap()
+    };
+    pr_debug!("observer.strong_count: {}", Arc::strong_count(observer));
+
+    observable.oob_wait.lock.raw_spin_unlock_irqrestore(flags);
+    Ok(nfr)
+}
+
+fn pull_notification<T: IoBufferWriter>(
+    observable: &mut RrosObservable,
+    observer: &mut Arc<RrosObserver>,
+    data: &mut T,
+    wait: bool,
+) -> Result<usize> {
+    let nfr_result: Result<Box<RrosNotificationRecord>>;
+    let nfr: Box<RrosNotificationRecord>;
+    let nf: __RrosNotification;
+    // let mut sigpoll: bool = false;
+    let flags: u64;
+
+    if running_inband().is_ok() {
+        // TODO:
+        nfr_result = pull_from_inband(observable, observer, wait);
+    } else {
+        nfr_result = pull_from_oob(observable, observer, wait);
+    }
+
+    if let Err(error) = nfr_result {
+        return Err(error);
+    }
+
+    nfr = nfr_result.unwrap();
+
+    nf = __RrosNotification {
+        tag: nfr.tag,
+        serial: nfr.serial,
+        issuer: nfr.issuer,
+        event: nfr.event.clone(),
+        date: ktime::ktime_to_timespec64(nfr.date),
+    };
+
+    let ret = unsafe {
+        data.write_raw(
+            &nf as *const _ as *const u8,
+            size_of::<__RrosNotification>(),
+        )
+    };
+
+    flags = observable.oob_wait.lock.raw_spin_lock_irqsave();
+
+    let ret = match ret {
+        Ok(_) => {
+            if observer.as_ref().pending_list.is_empty() {
+                observable.writable_observers += 1;
+            }
+
+            pr_debug!("observer.strong_count: {}", Arc::strong_count(observer));
+            unsafe {
+                Arc::get_mut_unchecked(observer).free_list.push_back(nfr);
+            }
+            pr_debug!("observer.strong_count: {}", Arc::strong_count(observer));
+
+            Ok(0_usize)
+        }
+        Err(error) => {
+            unsafe {
+                Arc::get_mut_unchecked(observer)
+                    .pending_list
+                    .push_front(nfr);
+            }
+            Err(error)
+        }
+    };
+
+    observable.oob_wait.lock.raw_spin_unlock_irqrestore(flags);
+
+    // TODO:
+    // if (unlikely(sigpoll))
+    // 	evl_signal_poll_events(&observable->poll_head,
+    // 			POLLOUT|POLLWRNORM);
+
+    ret
+}
+
+pub fn rros_read_observable<T: IoBufferWriter>(
+    observable: &mut RrosObservable,
+    data: &mut T,
+    count: usize,
+    wait: bool,
+) -> Result<usize> {
+    let sbr = dovetail::dovetail_current_state().subscriber() as *mut RrosSubscriber;
+    if sbr == 0 as *mut _ {
+        return Err(kernel::Error::ENXIO);
+    }
+    if !rros_element_is_observable(observable.element.clone()) {
+        return Err(kernel::Error::ENXIO);
+    }
+
+    let target = observable.element.clone().borrow().fundle;
+
+    let observer = unsafe { (*(*sbr).subscriptions.locked_data().get()).get_mut(&target) };
```

Comment:
It's strange to pass a `&<Arc` if you don't want to change it. Just dereference it and use `Arc` in the later functions.

Reply:
The `observer` is referenced in two places: one is the `RrosSubscriber` red-black tree, and the other is the `RrosObservable.observers` linked list. The `observer` changes when there are read and write operations on the observable. Therefore, it is necessary to call the `Arc::get_mut_unchecked(&mut self)` function to obtain a mutable reference, and the parameter of `Arc::get_mut_unchecked` is `&mut self`. Therefore, I believe that using `&Arc` here is necessary.